File: liriosimgcreate/_manifest.py
#
# This file is part of Liri.
#
# Copyright (C) 2019 Pier Luigi Fiorini <pierluigi.fiorini@gmail.com>
#
# SPDX-License-Identifier: GPL-3.0-or-later
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import humanfriendly
import yaml

__all__ = [
    'Manifest',
    'parse_manifest',
]

class Manifest(object):
    class Live(object):
        title = 'Linux'
        product = 'Linux (Live)'
        timeout = 600
        squashfs_compression = 'zstd'
    type = ''
    size = 4*1024*1024*1024
    osname = ''
    main_ref = ''
    refs = []
    remote_url = ''
    extra_kargs = []
    live = Live()


def parse_manifest(filename: str) -> Manifest:
    with open(filename, 'r') as fh:
        manifest = yaml.safe_load(fh)

    m = Manifest()
    m.type = manifest.get('type', m.type)
    m.size = humanfriendly.parse_size(manifest.get('size', str(m.size)))
    m.osname = manifest.get('osname', m.osname)
    m.main_ref = manifest.get('main-ref', m.main_ref)
    m.refs = [str(x) for x in manifest.get('refs', m.refs)]
    m.remote_url = manifest.get('remote-url', m.remote_url)
    m.extra_kargs = manifest.get('extra-kargs', m.extra_kargs)
    m.live.title = manifest.get('live', {}).get('title', m.live.title)
    m.live.product = manifest.get('live', {}).get('product', m.live.product)
    m.live.timeout = manifest.get('live', {}).get('timeout', m.live.timeout)
    m.live.squashfs_compression = manifest.get('live', {}).get('squashfs-compression', m.live.squashfs_compression)

    if m.type not in ('live', 'disk', 'qemu', 'vmware', 'metal'):
        raise SystemExit('Invalid image type')

    return m

File: src/creator/manifest.rs
/****************************************************************************
 * Copyright (C) 2020 Pier Luigi Fiorini <pierluigi.fiorini@gmail.com>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 ***************************************************************************/

use serde::Deserialize;
use serde_yaml;
use std::fs;
use std::path;

#[derive(Deserialize, Debug, PartialEq)]
pub enum ImageType {
    #[serde(rename = "disk")]
    Disk,

    #[serde(rename = "live")]
    Live,
}

#[derive(Deserialize, Default, Debug)]
pub struct LiveOptions {
    #[serde(default = "default_live_title")]
    pub title: String,

    #[serde(default = "default_live_product")]
    pub product: String,

    #[serde(default = "default_live_timeout")]
    pub timeout: u32,

    #[serde(rename(deserialize = "squashfs-compression"))]
    #[serde(default = "default_live_squashfs_compression")]
    pub squashfs_compression: String,
}

#[derive(Deserialize, Debug)]
pub struct Manifest {
    #[serde(rename(deserialize = "type"))]
    pub image_type: ImageType,

    #[serde(default = "default_size")]
    pub size: String,

    #[serde(default = "default_selinux")]
    pub selinux: bool,

    pub osname: String,

    #[serde(rename(deserialize = "main-ref"))]
    pub main_ref: String,

    #[serde(default)]
    pub refs: Vec<String>,

    #[serde(rename(deserialize = "remote-url"))]
    pub remote_url: String,

    #[serde(rename(deserialize = "extra-kargs"))]
    #[serde(default)]
    pub extra_kargs: Vec<String>,

    #[serde(default)]
    pub live: LiveOptions,
}

fn default_selinux() -> bool {
    false
}

fn default_size() -> String {
    "4 GiB".to_string()
}

fn default_live_title() -> String {
    "Linux".to_string()
}

fn default_live_product() -> String {
    "Linux (Live)".to_string()
}

fn default_live_timeout() -> u32 {
    600
}

fn default_live_squashfs_compression() -> String {
    "zstd".to_string()
}

impl Manifest {
    pub fn from_file(path: &path::Path, repo: Option<&str>) -> Result<Manifest, Box<dyn std::error::Error + 'static>> {
        let contents = fs::read_to_string(path)?;
        let mut manifest: Manifest = serde_yaml::from_str(&contents)?;
        if repo.is_some() {
            manifest.remote_url = String::from(format!("file://{}", repo.unwrap()));
        }
        return Ok(manifest);
    }
}
